File: hw/armijos/u3/hw29TemplateMethodStrategyComposite/TemplateMethod/TemplateMethodAplication/src/ec/edu/espe/composite/controller/CaffeineBeverage.java
package ec.edu.espe.composite.controller;

import javax.swing.JOptionPane;

/**
 *
 * @author Enzo Armijos, LogicLegion, DCCO-ESPE
 */
public abstract class CaffeineBeverage {

    public void prepareRecipe() {
        boilWater();
        brew();
        pourInCup();

    }

    void boilWater() {
        JOptionPane.showMessageDialog(null, "Boiling Water");
    }

    abstract void brew();

    public void pourInCup() {
        JOptionPane.showMessageDialog(null, "Pouring into cup");

    }
    abstract void addCondiments();
    boolean wantsCondiments(){
        return wantsCondiments();
    }
}

File: hw/villagomez/u3/hw29TemplateMethodAndComposite/TemplateMethodJS/model/CaffeineBeverage.js
export class CaffeineBeverage {
    prepareRecipe() {
        this.boilWater();
        this.brew();
        this.pourInCup();
        if (this.wantsCondiments()) {
            this.addCondiments();
        }
    }

    boilWater() {
        console.log("Boiling water");
    }

    brew() {
        throw new Error("This method should be overridden");
    }

    pourInCup() {
        console.log("Pouring into cup");
    }

    addCondiments() {
        throw new Error("This method should be overridden");
    }

    wantsCondiments() {
        throw new Error("This method should be overridden");
    }
}
